tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java

```diff
@@ -0,0 +1,359 @@
+            return data;
+        }
+
+        @Override
+        public void close() {
+            // nothing to do
+        }
+
+        @Override
+        public void onUpdate(ClusterResource clusterResource) {
+            if (expectClusterId) {
+                if (clusterResource.clusterId() == null) {
+                    throw new RuntimeException("expected cluster id to be " +
+                        "supported, but it was null.");
+                }
+            } else {
+                if (clusterResource.clusterId() != null) {
+                    throw new RuntimeException("expected cluster id to be " +
+                        "null, but it was supported.");
+                }
+            }
+        }
+    }
+
+    public void testConsume(final long prodTimeMs) throws Exception {
+        Properties consumerProps = new Properties();
+        consumerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, testConfig.bootstrapServer);
+        CompatibilityTestDeserializer deserializer = new CompatibilityTestDeserializer(testConfig.expectClusterId);
+        final KafkaConsumer<byte[], byte[]> consumer =
+                new KafkaConsumer<byte[], byte[]>(consumerProps, deserializer, deserializer);
+        final List<PartitionInfo> partitionInfos = consumer.partitionsFor(testConfig.topic);
+        if (partitionInfos.size() < 1)
+            throw new RuntimeException("expected at least one partition for topic " + testConfig.topic);
+        final Map<TopicPartition, Long> timestampsToSearch = new HashMap<>();
+        final LinkedList<TopicPartition> topicPartitions = new LinkedList<>();
+        for (PartitionInfo partitionInfo: partitionInfos) {
+            TopicPartition topicPartition = new TopicPartition(partitionInfo.topic(), partitionInfo.partition());
+            timestampsToSearch.put(topicPartition, prodTimeMs);
+            topicPartitions.add(topicPartition);
+        }
+        final OffsetsForTime offsetsForTime = new OffsetsForTime();
+        tryFeature("offsetsForTimes", testConfig.offsetsForTimesSupported,
+                new Runnable() {
+                    @Override
+                    public void run() {
+                        offsetsForTime.result = consumer.offsetsForTimes(timestampsToSearch);
+                    }
+                },
+                new Runnable() {
+                    @Override
+                    public void run() {
+                        log.info("offsetsForTime = {}", offsetsForTime.result);
+                    }
+                });
+        // Whether or not offsetsForTimes works, beginningOffsets and endOffsets
+        // should work.
+        consumer.beginningOffsets(timestampsToSearch.keySet());
+        consumer.endOffsets(timestampsToSearch.keySet());
+
+        consumer.assign(topicPartitions);
+        consumer.seekToBeginning(topicPartitions);
+        final Iterator<byte[]> iter = new Iterator<byte[]>() {
```

Comment:
Why do we need this? Can we not simply do a while loop in `fetchNext` and accumulate the results into a collection?